Verify compatible graph type

ldeaett opened the pull request:
Added function `verify_graph_is_compatible()` to make sure graph is simple (i.e., has no loops and is not a directed graph).  This function may disappear once support for these types of graphs is added, at which time new code will be required to apply search on the correct type of metagraph (meaning the subclass of metagraph corresponding to the type of primal graph that is passed in).

Commit:
Verify compatible graph type

alexhutman commented on src/zftools/metagraph.pyx:
```diff
@@ -31,14 +31,22 @@ from cpython.mem cimport (
+def verify_graph_is_compatible(sage_graph):
+    if sage_graph.is_directed():
+        raise ValueError("This is a directed graph.  Currently, directed graph zero forcing is not implemented.")
+    elif sage_graph.has_loops():
```
[nitpick] We can change this `elif` to `if` since we raise under the above condition

ldeaett replied:
That's true.  I kind of dislike `elif` for some reason.  Maybe because it's only one character away from being "elf" or something.  I'll change it.

alexhutman commented on src/zftools/metagraph.pyx:
```diff
@@ -31,14 +31,22 @@ from cpython.mem cimport (
                 )
-
+    
+    verify_graph_is_compatible(sage_graph)
```
It seems like we have 2 choices regarding where to put this:
1. Leave it in `zero_forcing_set()`, in which case I'd say let's move this check to the very top of the function
2. Put it at the beginning of `ZFSearchMetagraph.__init__()`

I say perhaps we can go with 1. since it's what the user calls, and we might have different requirements based on different forcing numbers. Thoughts?

ldeaett replied:
Hmm!!  Wow, you're right.  I think for sure 2 is the right way to go.  I wish I had thought of that.  It is true that there are different requirements for the primal graph as a function of what variant of zero forcing is being computed, but each instance of a metagraph (which, again, will be an abstract class at some point) computes a specific variant and therefore needs a specific type of graph.  So it really needs to be switched.  I'll get to work on that.

ldeaett replied:
Now I am going the other way on this.  Or maybe going both ways at once?

I realized that eventually the helper functions down the road are going to have to be able to instantiate different subclasses of the metagraph class according to what form of forcing is called for.  And decisions about whether the graph is directed or not (for example) need to be made in order to figure that out.  So this logic has to go in the helper functions.  But it also makes sense for each subclass of metagraph to be checking whether it is appropriate for the primal graph on which it is instantiated.  So... maybe doing both (1) and (2) makes sense?

Running into this issue is the only reason I have not yet pushed any new commit to address this.

For now, maybe the best solution is to implement (1) first, and later on maybe also require subclasses to provide (2) as a check as well.

alexhutman replied:
Hm I'm not too much of a fan of duplicating the requirements checking if we can avoid it. What do you think about doing (2), then in the helper functions, we can do something like:
```python
try:
    mg = ZFSearchMetagraph(some_graph)
except SomeException as e:
   # handle accordingly
```

ldeaett replied:
That's an interesting idea!  I had to think on this a bit, since it amounts to saying that if a certain subclass can be instantiated, then this implies that that subclass is in fact the appropriate one to use.  At first this felt like an assumption with some danger to it, but the more that I think about it, then more I think this might be right.  In short, absent any revelations I haven't had in the past 10 minutes, this does seem like the way to go.

alexhutman commented on src/zftools/metagraph.pyx:
```diff
@@ -31,14 +31,22 @@ from cpython.mem cimport (
 
+def verify_graph_is_compatible(sage_graph):
+    if sage_graph.is_directed():
+        raise ValueError("This is a directed graph.  Currently, directed graph zero forcing is not implemented.")
```
Let's change this to raise a `NotImplementedError` instead, for a bit more clarity. To differentiate the two exceptions, we can create 2 subclasses:
```python
# we can maybe make a new module called zftools.exceptions to house these?
class DirectedGraphError(NotImplementedError):
    pass

class LoopedGraphError(NotImplementedError):
    pass
```

ldeaett replied:
Cool.  I didn't even know about `NotImplementedError`!  I don't know about differentiating them.  That would require us to _a priori_ identify all possible reasons a metagraph subclass might decide that it can't instantiate on a certain graph.  But I guess we could just create an “other” subclass of `NotImplementedError`?  I guess the advantage to having something like `DirectedGraphError` is that then when `metagraph.__init__()` throws an error, the calling function (which is, say, `zero_forcing_number()`) can respond by printing some appropriate error message, or maybe also/instead branching according to which type of error was thrown so as to move on to a more appropriate subclass.

alexhutman commented on src/zftools/metagraph.pyx:
```diff
@@ -31,14 +31,22 @@ from cpython.mem cimport (
+    if sage_graph.is_directed():
+        raise ValueError("This is a directed graph.  Currently, directed graph zero forcing is not implemented.")
+    elif sage_graph.has_loops():
+        raise ValueError("This graph has loops.  Currently, looped zero forcing is not implemented.")
```
(see [above comment](https://github.com/alexhutman/zftools/pull/19/files#r1756210002))

Commit:
New exception handling code

alexhutman commented on src/zftools/metagraph.pyx:
```diff
@@ -31,15 +31,35 @@ from cpython.mem cimport (
+        frozenset start
+        frozenset end
+    
+    try:
```
Now that I see the end result, we should actually just let the exception bubble up here, that is, not do any try/catching. The reason for that is, as the code currently is, all we do is print a message. If someone calls this function from deep in some codebase, the print won't help them too much as the function call will just (likely, I haven't tested) do nothing and return `None`, and they might have to dig very deep to see why their code isn't working.

We should put those error messages in the exceptions themselves in `ZFSearchMetagraph.__init__` — `raise GraphIsDirectedError("This is a...")`. This way, the code becomes cleaner, and the user's code will crash if the exception is not caught, forcing them to handle it however they see fit.

ldeaett replied:
Your comment about what happens if “someone calls this function from deep in some codebase” makes total sense.  It is certainly possible that someone would use `zero_forcing_set()` to implement the computation of some other thing, at least in theory.  And so, while I was thinking of `zero_forcing_set()` as being _the_ interface to what we have working here, that is probably short-sighted.  All that is to say, we can probably do better than some dumb `print` statements.

At the same time, I think the reason not to _just_ “let the exception bubble up” is that `zero_forcing_set()` is the place where we should be able to instantiate different subclasses of the metagraph class according to what kind of forcing is called for.  We could either (a) have `zero_forcing_set()` itself sort out what kind of graph it is dealing with and throw exceptions if it can’t handle them or (b) code each subclass so that it can cough up an exception and then branch on why it’s complaining that it cannot handle forcing on the graph we are handing it.

Any ideas on which of (a) or (b) is better, or if there is an option (c) that I am not considering?

alexhutman replied:
I think (b) is the better option personally, because with (a), if someone were to manually create a `ZFSearchMetagraph` with a directed graph, they would presumably get an answer, but an incorrect one. (b) allows the class to have fine-grained control over what it can and cannot do. Sure, the user isn't _supposed_ to create it manually, but it's better to be safe than sorry IMO.

I suppose we can leave the `try/except` in `zero_forcing_set()` for now to reinforce that there will eventually be cases that will be handled. Regarding that though, do we know how exactly we're going to structure these new subclasses? Will we have a version of `ZFSearchMetagraph` that works on directed graphs? If so, I presume this will be its only functionality? If that is indeed the case, that means we will likely have to make 2 separate subclasses for directed graphs vs. undirected graphs for each new algorithm (i.e., the positive semidefinite ZF number), and 2 more for looped vs. non-looped graphs, right? If all of that sounds right, there must be a better way, as that doesn't really pass the sniff test for me.

Regarding what I meant for my previous comments, we're close, but let me push what I had in mind. You can certainly revert or write over it if you'd like also.

alexhutman replied:
In the commit I just pushed (https://github.com/alexhutman/zftools/pull/19/commits/e91a6e84377d6b7386235de6b31a06fca1f9c872), I just added the error message to the actual creation of the exception. I also removed the `try/catch` for now -- if you want, you can revert that part and do something like:
```python3
cdef:
    ZFSearchMetagraph metagraph
    frozenset start
    frozenset end
   
try:
    metagraph = ZFSearchMetagraph(sage_graph)
except GraphIsDirectedError:
    # This is where we will instantiate a metagraph for directed forcing
    raise
except GraphAllowsLoopsError:
    # This is where we will instantiate a metagraph for (undirected) looped forcing
    raise
start = frozenset()
end = frozenset(metagraph.to_relabeled_metavertex(sage_graph.vertices(sort=False)))
return metagraph.dijkstra(start, end)
```

In my opinion, this is a little unnecessary since we can just add this once we get closer to actually handling these exceptions via the new subclasses, but I wouldn't be opposed to it. [Here](https://docs.python.org/3/tutorial/errors.html#raising-exceptions) is a section from the Python docs for reference. Note the bottom of the last example in the section: we can call `raise` (with no expression afterwards) inside of an `except` clause in order to re-raise the caught exception.

ldeaett committed:
Removed print statements at least

alexhutman committed:
[metagraph.pyx] Include error msg in raised exceptions

Also removed try/catch. We can add it back once we're closer with the subclasses.
